### ISSUE TITLE
Update stack package command to support templating in stacks

### ISSUE THREAD

**Issue by neeraj-laad** [appsody/appsody#560]
This will be follow on work from the exploratory work done under https://github.com/appsody/stacks/issues/430

We now need to:

- [ ] Agree on what information from stack.yaml will be available to stack creators
- [ ] Agree on keys/names used for the template variable
- [ ] Agree how will users specify custom template strings
- [ ] Update stack package command to do this template parsing/replacement before building the stack.
- [ ] Validate this works on all platforms and retains file permissions

**Comment by neeraj-laad** [comment]
We need the following variables as a minimum:

- StackID
- StackName
- SackMajorVersion
- StackMinorVersion
- StackPatchVersion
- StackImageNamespace

**Comment by kylegc** [comment]
Lots of this information overlaps with the image labels we have added. We don't necessarily need to use the same variable names, but we should reuse code and values as much as possible. For instance, we calculate a created timestamp for the labels, if we add that as a template variable, it should use the **exact same** timestamp, not one that is off by even a couple milliseconds.

**Comment by kylegc** [comment]
I definitely think we should have a way for the stack developer to define their own arbitrary template variable=value pairs for replacement. These would likely go in the stack.yaml under a defined array. 
We could consider adding these to the image labels as well.

**Comment by kylegc** [comment]
If the stack developer is using templating within the stack, I don't like making them wrap their templates with `{{"{{.UserVariableName}}"}}`. That could become very cumbersome for the stack developer.

Thus, I think we should use a different delimiter for the stack variables, but one similar enough that it still looks like a template. Here are some ideas:
- `{{{.StackId}}}`
- `{{$.StackId}}`  - yes I do believe the starting delimiter can be different from the ending
- `{{appsody.StackId}}`

**Comment by neeraj-laad** [comment]
@kylegc @CameronMcWilliam I've been thinking more on this and have discussed this with a few others to see what they find intuitive. 

The general feedback I'm getting at the moment is that users would prefer to use something that is simple and standard. That way they will know exactly how to work around limitations like escaping. 

Most did like {{stack`.id`}} as it looks meaningful to user and could provide some scoping if we wanted to provide templating from other files in future. But there are 2 big negatives too:
- IDEs might format `{{ stack.id }}` with spaces and that might break things.
- Also using other go templating capabilities like logic etc will be cumberson.

We do not know when and if we'll see this overlap in our stacks. So, keeping it simple for now makes sense to me. So I'm proposing we stick to `{{` and `}}` as delimiters. 

If / when we get to the conflict with multiple templating engines, we might be able to solve that problem in a different way:
- let stack creator specify their own delimiters
- let then exclude/ include certain files from templating.

**Comment by neeraj-laad** [comment]
So, I'm thinking of these: 

```
{{ .stack.id }}
{{ .stack.name }}
{{ .stack.description }}
{{ .stack.created }}
{{.stack.image.namespace }}
{{ .stack.version.semver }} or {{ .stack.version.full }}
{{ .stack.version.major }}
{{ .stack.version.minor }}
{{ .stack.versoin.patch }}
{{ .stack.key1 }}
{{ .stack.key2 }}
```

Where key1 and key2 is user-specified templating-data in `stack.yaml`.

```
templating-data:
  key1: val1
  key2: val2
```

Open to suggestions to better term for `templating-data`.

**Comment by kylegc** [comment]
Hi Neeraj, all valid points. I'm good with that strategy and those variable names. My main goal was to make the templating look and feel normal as well.

**Comment by kylegc** [comment]
I have two additional variables to add:
- `.stack.tag` This will be the full tag of the image, such as `appsody/nodejs-express:0.2.7` Although you could construct this from the other variables, we do have a label for this and I could see being a useful shortcut.
- `.stack.version.majorminor` This would be a combination of just the major and the minor `0.2` and again a useful shortcut.

I also think the full version should stay just `.stack.version` to be consistent with the labels

**Comment by CameronMcWilliam** [comment]
I like the addition of these two variables, however we can't call the full version number `.stack.version` as `version` is a map holding all the smaller values so if you call `.stack.version` you will get the whole map returned.  Welcome to ideas of a better name for `versionFull` though.

**Comment by neeraj-laad** [comment]
@kylegc I agree just calling it version would be better.. but that seemed difficult as version is now a map not a string. Perhaps something here will give a clearer name - https://semver.org/#backusnaur-form-grammar-for-valid-semver-versions

I like the major minor shortcut too, will be handy for certain cases

One thing I noticed will be useful for swift and kitura is maintainer (they are looking to override description, maintainer tags from base image and they could use templating to keep in sync with stack.yaml. Could we construct a meaningful string for that perhaps list of emails separated by comma?

 Also would `stack.image` be a better name for .`stack.tag`? Not against tag either if that is consistent with labels.

**Comment by CameronMcWilliam** [comment]
Maybe we do call it `{{.stack.version.core}}` to keep in line with SemVer

**Comment by kylegc** [comment]
We have mantainers as a comma separated sting in the labels, so Cameron can simply use that :-)

**Comment by kylegc** [comment]
Bummer we can't use `{{ .stack.version }}`. That is probably the most common use and it matches the labels.
What about this?
```
{{ .stack.version }}
{{ .stack.semver.major }}
{{ .stack.semver.minor }}
{{ .stack.semver.patch }}
{{ .stack.semver.majorminor }}
```

**Comment by CameronMcWilliam** [comment]
Yep found those labels :) Also I like this `semver` syntax, I agree `version.full` is a bit odd and this seems like a suitable workaround, I'll see what @neeraj-laad thinks.

**Comment by CameronMcWilliam** [comment]
After discussion with @neeraj-laad we agreed your version variables are more intuitive so we're changing them.  Something else to note, as there's a lot of cross over from the labels, but the names of these labels don't match the `stack.yaml` variables (e.g. `name` in `stack.yaml` is `title` in labels), for now we're going to call the variables what they are called in the `stack.yaml` as this is what stack creators will use to apply templating.  Whether at some point we want to match labels, `stack.yaml` and templating variables, that can be fixed in a later iteration.

**Comment by kylegc** [comment]
I agree with this approach. Thanks!

**Comment by neeraj-laad** [comment]
Closing this as this has now been implemented and released in CLI v0.5.0
